elytron/src/main/java/org/wildfly/extension/elytron/PolicyDefinitions.java

```diff
@@ -323,9 +349,17 @@ public void accept(Consumer<Policy> policyConsumer) {
                         Policy policy = newPolicy(policyProvider, configuredClassLoader);
                         policyConsumer.accept(policy);
 
-                        doPrivileged((PrivilegedExceptionAction<PolicyConfigurationFactory>) () -> newPolicyConfigurationFactory(
-                                configurationFactory,
-                                defaultConfigurationFactory ? PolicyDefinitions.class.getClassLoader() : configuredClassLoader));
+                        if (context.isBooting() && legacyJacc) {
+                            throw ElytronSubsystemMessages.ROOT_LOGGER.unableToEnableJaccSupport();
+                        }
+
+                        if (legacyJacc) {
+                            context.restartRequired();
```

Comment:
This feels wrong.  Not that context.restartRequired is called somewhere, but that it is called here.

Code that changes the runtime and then calls either reloadRequired or restartRequired is likely buggy code. Those things are logical alternatives.

This (and the exception throwing above) seems like something that should be handled in PolicyDefinitions.getPolicy() in its performRuntime implementation.

I also question whether the restartRequired is needed.  I don't think the 'context.isBooting()' at L352 is quite right.  Either the config is valid or it is not and the op that made it invalid should be rejected. You don't want to tell them to restart, and then they do and the boot fails.  The use case for not failing immediately would be so they could add a second op that removes legacy JACC. But if two ops are required to make a change, and the first leaves things in an invalid state, then the user should combine the two ops in a composite (e.g. a CLI batch). If the user used a composite like they should, when this OperationContext.Stage.RUNTIME handling executes the org.wildfly.legacy-security.jacc would no longer be present.

The one case where we do allow invalid sets of capabilities is during boot in an admin-only server. So that combination could WARN and set restartRequired. Otherwise the context should be setRollbackOnly and an OperationFailedException thrown.  And the 'boot in an admin-only server' case only needs to be considered if the OSH actually does runtime work in admin-only. (Most subsystem OSHs don't, but Elytron is unusual in this regard.)

Reply:
The problem I was trying to solve with the context.isBooting was the sequence of 
```
/subsystem=security:write-attribute(name=initialize-jacc, value=false)
/subsystem=elytron/policy=jacc:add(jacc-policy={})
```
This leads to error since org.wildfly.legacy-security.jacc is not de-regestired until restart. I guess a better solution would be to de-register the capability in similar way Policy remove handler is doing? Or is OK to fail the 2nd operation in this case until the server is restarted?

Reply:
It is fine to fail. A user should use a composite / CLI batch for this kind of change.

Reply:
@bstansberry a composite wouldn't work here - a problem with the JACC APIs is they don't provide a way to replace the policy after first initialisation.

A more complex solution would be we register some form of delegate policy where we can subsequently swap in a different instance - or look to contribute changes to the APIs to allow a replacement but there are probably a lot of expectations out there about the stability of the global policy.

Reply:
@darranl could we  make the elytron policy changes be applied to runtime only during boot? This way we could guarantee the policy is only initialised once.

Reply:
I don't believe this is about initialisation once, we want to ensure it is initialised to the correct value and trigger a restart if it is not the correct value.

Reply:
Got it; sorry I didn't understand the issue properly before and replied hastily. :( 

So AIUI there are two kinds of problems related to the use of capabilities here.

One is that a config change to legacy security can result in the "org.wildfly.legacy-security.jacc" capability being gone from the config, but the Policy is still in place. And then this handler needs to decide what to do. That's the more fundamental problem that I'll focus on below.

The other is that the initialize-jacc attribute's value is what really determines whether that capability should be present. The handling of that is an issue in the legacy security subsystem, but it's better to discuss that there.

Re the more fundamental problem the only idea I have is something we called a 'tombstone' in some old clustering code -- a marker for something that used to be alive but is now gone. So a "org.wildfly.legacy-security.jacc.tombstone" capability. It gets added when "org.wildfly.legacy-security.jacc" is added but never gets removed (except of course by process exit).

Presence of "org.wildfly.legacy-security.jacc" when Stage.RUNTIME executes for this add handler is an error condition (except maybe during admin-only boot.) But presence of "org.wildfly.legacy-security.jacc.tombstone" is not an error condition, it's just a sign that the Elytron JACC cannot be put in place now and restart is required.

Perhaps at some point the general capability system could add some kind of less hacky support for this general concept.

Reply:
I think we do get to keep in mind one of the two subsystems registering this is deprecated and destined to be deleted at some point.